p2p/server/server.go

```diff
@@ -235,44 +235,33 @@ func (s *Server) queueHandler(ctx context.Context, stream network.Stream) {
 
 // Request sends a binary request to the peer. Request is executed in the background, one of the callbacks
 // is guaranteed to be called on success/error.
-func (s *Server) Request(
-	ctx context.Context,
-	pid peer.ID,
-	req []byte,
-	resp func([]byte),
-	failure func(error),
-) error {
+func (s *Server) Request(ctx context.Context, pid peer.ID, req []byte) ([]byte, error) {
 	start := time.Now()
 	if len(req) > s.requestLimit {
-		return fmt.Errorf("request length (%d) is longer than limit %d", len(req), s.requestLimit)
+		return nil, fmt.Errorf("request length (%d) is longer than limit %d", len(req), s.requestLimit)
 	}
 	if s.h.Network().Connectedness(pid) != network.Connected {
-		return fmt.Errorf("%w: %s", ErrNotConnected, pid)
+		return nil, fmt.Errorf("%w: %s", ErrNotConnected, pid)
 	}
-	go func() {
-		data, err := s.request(ctx, pid, req)
-		if err != nil {
-			failure(err)
-		} else if len(data.Error) > 0 {
-			failure(errors.New(data.Error))
-		} else {
-			resp(data.Data)
-		}
-		s.logger.WithContext(ctx).With().Debug("request execution time",
-			log.String("protocol", s.protocol),
-			log.Duration("duration", time.Since(start)),
-			log.Err(err),
-		)
-		switch {
-		case s.metrics == nil:
-			return
-		case err != nil:
-			s.metrics.clientLatencyFailure.Observe(time.Since(start).Seconds())
-		case err == nil:
-			s.metrics.clientLatency.Observe(time.Since(start).Seconds())
-		}
-	}()
-	return nil
+	data, err := s.request(ctx, pid, req)
+	if err != nil {
+		return nil, err
+	} else if len(data.Error) > 0 {
+		return nil, errors.New(data.Error)
+	}
+	s.logger.WithContext(ctx).With().Debug("request execution time",
+		log.String("protocol", s.protocol),
+		log.Duration("duration", time.Since(start)),
+		log.Err(err),
+	)
+	switch {
+	case s.metrics == nil:
+	case err != nil:
+		s.metrics.clientLatencyFailure.Observe(time.Since(start).Seconds())
+	case err == nil:
+		s.metrics.clientLatency.Observe(time.Since(start).Seconds())
+	}
+	return data.Data, nil
```

Comment:
The metrics won't observe failures, since the function returns on an error before metrics are reported. I suggest changing it to this:

```suggestion
	data, err := s.request(ctx, pid, req)
	s.logger.WithContext(ctx).With().Debug("request execution time",
		log.String("protocol", s.protocol),
		log.Duration("duration", time.Since(start)),
		log.Err(err),
	)
	switch {
	case s.metrics == nil:
	case err != nil:
		s.metrics.clientLatencyFailure.Observe(time.Since(start).Seconds())
		return nil, err
	case len(data.Error) > 0:
		s.metrics.clientLatencyFailure.Observe(time.Since(start).Seconds())
		return nil, errors.New(data.Error)
	default:
		s.metrics.clientLatency.Observe(time.Since(start).Seconds())
		return data.Data, nil
	}
```

The first 2 early exits (request too large and network not connected don't need to be captured by metrics in my opinion since no request is actually made 🙂